### PR TITLE
Implement versioning for CI/CD workflow.

### LINKED ISSUE #1182: Proposal: Use semantic versioning
# Proposal

Looking into the [tags](https://hub.docker.com/r/mailu/admin/tags), I realized, that you always push into the the main tags 1.6 and 1.7. Like this, it is hard to determine, whether my installed version has an update or not.

Thus, with [semantic versioning](https://semver.org/) it is easier to check, whether there is a new version, e.g., an update, available. Especially for automatic tools. Also if an update fails, it's possible to rollback to the previous version (of course with a backup).

As a reference, i can recommend mysql. They do a great job with versioning:

* https://hub.docker.com/_/mysql?tab=description

# Examples

* I have a working 1.6 and I update to a broken 1.6. There is no way to roll back to the working 1.6.
* If have a working 1.6.1 and I update to a broken 1.6.2. Like this, there is an easy way roll back to the working 1.6.1

### PR DESCRIPTION
**Pull request by Diman0**
## What type of PR?

Feature!

## What does this PR do?
This PR introduces 3 things
- Add versioning (tagging) for branch x.y (1.8). E.g. 1.8.0, 1.8.1 etc.
  - docker repo will contain x.y (latest) and x.y.z (pinned version) images.
  - The X.Y.Z tag is incremented automatically. E.g. if 1.8.0 already exists, then the next merge on 1.8 will result in the new tag 1.8.1 being used.
- Make the version available in the image.
  -  For X.Y and X.Y.Z write the version (X.Y.Z) into /version on the image and add a label with version=X.Y.Z
	  -  This means that the latest X.Y image shows the pinned version (X.Y.Z e.g. 1.8.1) it was based on. Via the tag X.Y.Z you can see the commit hash that triggered the built.
  -  For master write the commit hash into /version on the image and add a label with version={commit hash}
-  Automatic releases. For x.y triggered builts (e.g. merge on 1.9) do a new github release for the pinned x.y.z (e.g. 1.9.2). 
  -  Release shows a static message (see RELEASE_TEMPLATE.md) that explains how to reach the newsfragments folder and change the branch to the tag (x.y.z) mentioned in the release. Now you can get the changelog by reading all newsfragment files in this folder.

This PR does not change anything to our workflow (what we (human persons) do). Our processes are still exactly the same. The above introduced logic is automatic. When we backport to X.Y all the magic for creating the pinned version X.Y.Z is handled by the CI/CD workflow.

### Related issue(s)
- closes #1182

## Prerequisites
Before we can consider review and merge, please make sure the following list is done and checked.
If an entry in not applicable, you can check it or remove it from the list.

- [x] In case of feature or enhancement: documentation updated accordingly
- [x] Unless it's docs or a minor change: add [changelog](https://mailu.io/master/contributors/workflow.html#changelog) entry file.

## Testing
Suggested testing steps. This should cover all situations including BORS. It does require that you use your own docker repo or temporarily create a new one.
Suggested testing steps.
1. Create new github repo.
2. Add the required docker secrets to the project (see beginning of CI.yml for the secret names), DOCKER_UN, DOCKER_PW, DOCKER_ORG, DOCKER_ORG_TESTS.
3. Clone the project.
4. Copy the contents of the PR to the cloned project.
5. Push to your new github repo.
6. Now master images are built. Check that images with tag master are pushed to your docker repo
7. Check with docker inspect nginx:master that it has the label version={commit hash}.
8. Run an image, run `docker-compose exec <name> cat /version`. Note that /version also contains the pinned version. For master the pinned version is the commit hash.
9. Create branch 1.8. 
10. Push branch 1.8 to repo.
11. Note that tags 1.8 and 1.8.0 are built and pushed to docker repo
12. Inspect label and /version. Note that 1.8 and 1.8.0 both show version 1.8.0.
13. Push another commit to branch 1.8.
14. Note that tags 1.8 and 1.8.1 are built and pushed to docker repo
15. Inspect label and /version. Note that 1.8 and 1.8.1 both show version 1.8.1.
16. Let's check BORS stuff.
17. Create branch testing.
18. Push the commit with the exact commit text (IMPORTANT!!): `Try #1234:`'.
19. Note that images are built and pushed for tag `pr-1234`.
20. Inspect label and /version. Note that the version is `pr-1234`.
20. Create branch staging.
21. Push the commit with commit text: `Merge #1234`.
22. Note that this image is not pushed to docker (as expected).

but you could also check the GH repo and docker repo I used:
https://github.com/Diman0/Mailu_Fork
https://hub.docker.com/r/diman/rainloop/tags

### CODE REVIEW

**Comment by Diman0** [comment]
bors try

**Comment by Diman0** [comment]
bors try-

**Comment by Diman0** [comment]
bors try

**Comment by Diman0** [comment]
The first CI workflow refused to start because I forgot to whitelist the new action. All actions in a workflow must be whitelisted before they can be used.

File: webmails/roundcube/Dockerfile [hunk 1]
```diff
@@ -7,9 +7,12 @@ ONBUILD COPY --from=balenalib/rpi-alpine:3.14 /usr/bin/qemu-arm-static /usr/bin/
 FROM ${ARCH}php:7.4-apache as build_other
 
 FROM build_${QEMU}
-
+ARG VERSION
 ENV TZ Etc/UTC
 
+LABEL version=$VERSION
+RUN echo $VERSION >> /version
```
Comment (ghostwheel42) [inline webmails/roundcube/Dockerfile:14]: I think this should be the last "RUN" command in the Dockerfiles's to allow using unmodified / cached states.

File: .github/workflows/CI.yml [hunk 2]
```diff
@@ -28,34 +24,68 @@ on:
 #    The docker repository for test images. Only used for the branch TESTING (BORS try).
 # Add the above secrets to your github repo to determine where the images will be pushed.
 ################################################
+# Code block that is used as one liner.
+##!/bin/bash
+#version=$( git tag --list "{{ env.MAILU_VERSION }}.*" | tail -1  )
+#root_version=${version:0:3}
+#patch_version=${version:4:4}
```
Comment (ghostwheel42) [inline .github/workflows/CI.yml:31]: This does not account for versions > 9.
Maybe you can use
```
root_version=${version%.*}
patch_version=${version##*.}
```

File: .github/workflows/CI.yml [hunk 3]
```diff
@@ -28,34 +24,68 @@ on:
...
+        run: |
+          echo "MAILU_VERSION=${{ env.BRANCH }}" >> $GITHUB_ENV
+          echo "DOCKER_ORG=$DOCKER_ORG" >> $GITHUB_ENV
+      - name: Derive PINNED_MAILU_VERSION for normal release x.y
+        if: ${{ env.BRANCH != 'testing' && env.BRANCH != 'staging' && env.BRANCH != 'master' }}
+        shell: bash
         run: |
-          echo "MAILU_VERSION=${{ env.MAILU_BRANCH }}" >> $GITHUB_ENV
-          echo "DOCKER_ORG=${{ secrets.DOCKER_ORG }}" >> $GITHUB_ENV
+          version=$( git tag --list "${{ env.MAILU_VERSION }}.*" | tail -1  );root_version=${version:0:3};patch_version=${version:4:4};if [ "$patch_version" == "" ]; then pinned_version=${{ env.MAILU_VERSION }}.0; else pinned_version=$root_version.$(expr $patch_version + 1); fi;echo "PINNED_MAILU_VERSION=$pinned_version" >> $GITHUB_ENV
```
Comment (ghostwheel42) [inline .github/workflows/CI.yml:76]: see above

File: .github/workflows/CI.yml [hunk 4]
```diff
@@ -89,26 +119,49 @@ jobs:
...
-          MAILU_BRANCH: ${{ env.BRANCH }}
+          DOCKER_ORG: ${{ secrets.DOCKER_ORG }}
+        run: |
+          echo "MAILU_VERSION=${{ env.BRANCH }}" >> $GITHUB_ENV
+          echo "DOCKER_ORG=$DOCKER_ORG" >> $GITHUB_ENV
+      - name: Derive PINNED_MAILU_VERSION for normal release x.y
+        if: ${{ env.BRANCH != 'testing' && env.BRANCH != 'master' }}
+        shell: bash
+        run: |
+          version=$( git tag --list "${{ env.MAILU_VERSION }}.*" | tail -1  );root_version=${version:0:3};patch_version=${version:4:4};if [ "$patch_version" == "" ]; then pinned_version=${{ env.MAILU_VERSION }}.0; else pinned_version=$root_version.$(expr $patch_version + 1); fi;echo "PINNED_MAILU_VERSION=$pinned_version" >> $GITHUB_ENV
```
Comment (ghostwheel42) [inline .github/workflows/CI.yml:152]: see above

File: .github/workflows/CI.yml [hunk 5]
```diff
@@ -138,26 +191,49 @@ jobs:
...
```
Comment (ghostwheel42) [inline .github/workflows/CI.yml:224]: see above

File: .github/workflows/CI.yml [hunk 6]
```diff
@@ -187,26 +263,49 @@ jobs:
...
```
Comment (ghostwheel42) [inline .github/workflows/CI.yml:296]: see above

File: .github/workflows/CI.yml [hunk 7]
```diff
@@ -236,26 +335,49 @@ jobs:
...
```
Comment (ghostwheel42) [inline .github/workflows/CI.yml:368]: see above

File: .github/workflows/CI.yml [hunk 8]
```diff
@@ -285,26 +407,49 @@ jobs:
...
```
Comment (ghostwheel42) [inline .github/workflows/CI.yml:440]: see above

File: .github/workflows/CI.yml [hunk 9]
```diff
@@ -334,26 +479,49 @@ jobs:
...
         run: |
-          echo "MAILU_VERSION=${{ env.MAILU_BRANCH }}" >> $GITHUB_ENV
-          echo "DOCKER_ORG=${{ secrets.DOCKER_ORG }}" >> $GITHUB_ENV
+          echo "MAILU_VERSION=${{ env.BRANCH }}" >> $GITHUB_ENV
+          echo "DOCKER_ORG=$DOCKER_ORG" >> $GITHUB_ENV
+      - name: Derive PINNED_MAILU_VERSION for normal release x.y
+        if: ${{ env.BRANCH != 'testing' && env.BRANCH != 'master' }}
+        shell: bash
+        run: |
+          version=$( git tag --list "${{ env.MAILU_VERSION }}.*" | tail -1  );root_version=${version:0:3};patch_version=${version:4:4};if [ "$patch_version" == "" ]; then pinned_version=${{ env.MAILU_VERSION }}.0; else pinned_version=$root_version.$(expr $patch_version + 1); fi;echo "PINNED_MAILU_VERSION=$pinned_version" >> $GITHUB_ENV
```
Comment (ghostwheel42) [inline .github/workflows/CI.yml:512]: see above

File: .github/workflows/CI.yml [hunk 10]
```diff
@@ -389,27 +557,49 @@ jobs:
...
```
Comment (ghostwheel42) [inline .github/workflows/CI.yml:590]: see above

File: .github/workflows/CI.yml [hunk 11]
```diff
@@ -430,13 +620,49 @@ jobs:
...
+        env:
+          DOCKER_ORG: ${{ secrets.DOCKER_ORG }}
+        run: |
+          echo "MAILU_VERSION=${{ env.BRANCH }}" >> $GITHUB_ENV
+          echo "DOCKER_ORG=$DOCKER_ORG" >> $GITHUB_ENV
+      - name: Derive PINNED_MAILU_VERSION for normal release x.y
+        if: ${{ env.BRANCH != 'testing' && env.BRANCH != 'staging' && env.BRANCH != 'master' }}
+        shell: bash
+        run: |
+          version=$( git tag --list "${{ env.MAILU_VERSION }}.*" | tail -1  );root_version=${version:0:3};patch_version=${version:4:4};if [ "$patch_version" == "" ]; then pinned_version=${{ env.MAILU_VERSION }}.0; else pinned_version=$root_version.$(expr $patch_version + 1); fi;echo "PINNED_MAILU_VERSION=$pinned_version" >> $GITHUB_ENV
```
Comment (ghostwheel42) [inline .github/workflows/CI.yml:652]: see above

File: tests/deploy.sh [hunk 12]
```diff
@@ -1,7 +1,51 @@
...
+  exit 0
+fi
+
+#Deploy for main releases
+#Images are built with tag PINNED_MAILU_VERSION (x.y.z).
+#We are tagging them as well with MAILU_VERSION (x.y)
+#After that, both tags are pushed to the docker repository.
+if [ "$PINNED_MAILU_VERSION" != "" ] && [ "$BRANCH" != "master" ]
+then
+  images=$(docker-compose -f tests/build.yml config | grep 'image: ' | awk -F ':' '{ print $2 }')
```
Comment (ghostwheel42) [inline tests/deploy.sh:20]: you don't need an extra grep. awk can do this like this:
```
awk -F ':' '/image:/{ print $2 }'
```

File: tests/deploy.sh [hunk 13]
```diff
@@ -1,7 +1,51 @@
...
+  exit 0
+fi
+
+#Deploy for master. For master we only publish images with tag master
+#Images are built with tag PINNED_MAILU_VERSION (commit hash).
+#We are tagging them as well with MAILU_VERSION (master)
+#Then we publish the images with tag master
+if [ "$PINNED_MAILU_VERSION" != "" ] && [ "$BRANCH" == "master" ]
+then
+  images=$(docker-compose -f tests/build.yml config | grep 'image: ' | awk -F ':' '{ print $2 }')
```
Comment (ghostwheel42) [inline tests/deploy.sh:39]: see above

**Review by ghostwheel42 — CHANGES_REQUESTED** [review]
Hi @Diman0 - thanks for this PR.
I haven't tested it yet, but here a a few comments.

File: webmails/roundcube/Dockerfile (hunk 1 above)
Comment (Diman0) [reply webmails/roundcube/Dockerfile:14]: Yes. That makes much sense. I will make this change.

File: .github/workflows/CI.yml [hunk 14]
```diff
@@ -9,10 +9,6 @@ on:
       - '1.7'
       - '1.8'
       - master
-      # version tags, e.g. 1.7.1
-      - '[1-9].[0-9].[0-9]'
-      # pre-releases, e.g. 1.8-pre1
-      - 1.8-pre[0-9]
```
[the remark this answers is not in the document]
Comment (Diman0) [reply .github/workflows/CI.yml:15]: We do not make use of x.y.z and 1.8-pre branches. That is why I removed it. I also prefer that we have to add explicitly any new branches for triggers of workflow runs. This gives us more control.

File: tests/deploy.sh (hunk 12 above)
Comment (Diman0) [reply tests/deploy.sh:20]: I will give it a try. It is better to have less piping if possible.

File: .github/workflows/CI.yml (hunk 14 above)
[the remark this answers is not in the document]
Comment (ghostwheel42) [reply .github/workflows/CI.yml:15]: Oh okay, my fault.
I assumed it was added.

**Comment by Diman0** [comment]
bors try

**Comment by Diman0** [comment]
bors try

**Comment by Diman0** [comment]
@nextgens I added you to my cloned repo https://github.com/Diman0/Mailu_Fork . I'm not sure if you have all the necessary access. I cannot configure any access permissions yet until you have accepted the repo.
The secrets are configured so that testing images and merged images are pushed to repo https://hub.docker.com/u/diman.

You could start testing from step 9 in my test plan, except for that you enhance the CI.yml to allow branch 1.9 and push to branch 1.9.

**Review by nextgens — DISMISSED** [review]
LGTM

The only edge case I've found is when a new commit is pushed on a branch while the previous build hasn't finished; bors is supposed to prevent concurrent workflow execution for stuff on the same branch.

**Comment by Diman0** [comment]
bors try

**Review by nextgens — APPROVED** [review]
